File: leetcode/last-stone-weight.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    int lastStoneWeight(vector<int>& stones)
    {
        int one, two;
        make_heap(stones.begin(), stones.end());

        while (!stones.empty()) {
            if (stones.size() == 1) {
                return stones.front();
            }
            one = stones.front();
            pop_heap(stones.begin(), stones.end());
            stones.pop_back();

            two = stones.front();
            pop_heap(stones.begin(), stones.end());
            stones.pop_back();

            if (one != two) {
                stones.push_back(one - two);
                push_heap(stones.begin(), stones.end());
            }
        }

        return 0;
    }
};

File: leetcode/last-stone-weight.py
from heapq import heapify, heappop, heappush
from typing import List


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones = [-s for s in stones]

        heapify(stones)
        while stones:
            if len(stones) == 1:
                return -stones[0]

            one = heappop(stones)
            two = heappop(stones)

            if two != one:
                heappush(stones, one - two)

        return 0
